### PR TITLE
CVSL-1860 add review without variation events into timeline

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: server/data/licenceApiClient.ts [hunk 1]
```diff
@@ -545,10 +545,13 @@ export default class LicenceApiClient extends RestClient {
...
-  async reviewWithoutVariation(licenceId: number): Promise<void> {
-    await this.post({
-      path: `/licence/id/${licenceId}/review-with-no-variation-required`,
-    })
+  async reviewWithoutVariation(licenceId: number, user: User): Promise<void> {
+    await this.post(
+      {
+        path: `/licence/id/${licenceId}/review-with-no-variation-required`,
+      },
+      { username: user.username }
```
Comment (andrewrlee) [inline server/data/licenceApiClient.ts:553]: adding missing user so audit occurs correctly

File: server/enumeration/licenceEventType.ts [hunk 2]
```diff
@@ -9,6 +9,8 @@ enum LicenceEventType {
   VARIATION_SUBMITTED = 'VARIATION_SUBMITTED',
   VARIATION_REFERRED = 'VARIATION_REFERRED',
   VARIATION_APPROVED = 'VARIATION_APPROVED',
+  HARD_STOP_REVIEWED_WITHOUT_VARIATION = 'HARD_STOP_REVIEWED_WITHOUT_VARIATION',
+  HARD_STOP_REVIEWED_WITH_VARIATION = 'HARD_STOP_REVIEWED_WITH_VARIATION',
```
Comment (andrewrlee) [inline server/enumeration/licenceEventType.ts:13]: we're not displaying this but it is a valid event type now

File: server/routes/varyingLicences/handlers/timeline.ts [hunk 3]
```diff
@@ -1,9 +1,13 @@
 import { Request, Response } from 'express'
 import LicenceStatus from '../../../enumeration/licenceStatus'
+import TimelineService from '../../../services/timelineService'
 import LicenceService from '../../../services/licenceService'
 
 export default class TimelineRoutes {
-  constructor(private readonly licenceService: LicenceService) {}
+  constructor(
+    private readonly licenceService: LicenceService,
+    private readonly timelineService: TimelineService
```
Comment (andrewrlee) [inline server/routes/varyingLicences/handlers/timeline.ts:9]: split out new service for creating timeline

File: server/services/timelineService.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        new TimelineEvent(
+          TimelineEventType.REVIEWED_WITHOUT_VARIATION,
+          'Licence reviewed without being varied',
+          licence.statusCode,
+          convertToTitleCase(`${e.forenames} ${e.surname}`),
+          licence.id,
+          e.eventTime
+        )
+    )
+  }
```
Comment (andrewrlee) [inline server/services/timelineService.ts:36]: new function to create timeline events from licence events (there will only ever be one per licence)

These licence events are then concat-ed to the standard time line events and the whole lot is sorted by time, most recent first

(the standard time line events are a bit wrong and misleading - haven't fixed those)

File: server/services/timelineService.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+          licence.statusCode,
+          creator,
+          licence.id,
+          licence.dateLastUpdated
+        )
+
+      case LicenceStatus.ACTIVE:
+      case LicenceStatus.INACTIVE:
+      default: {
+        const prisonName = convertToTitleCase(licence.prisonDescription).replace('hmp', 'HMP').replace('yoi', 'YOI')
```
Comment (andrewrlee) [inline server/services/timelineService.ts:127]: handling weird abbreviations in prison type

File: server/services/timelineService.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+              licence.statusCode,
+              creator,
+              licence.id,
+              licence.dateLastUpdated
+            )
+          : new TimelineEvent(
+              TimelineEventType.CREATION,
+              'Licence created',
+              licence.statusCode,
+              `${creator}${licence.kind === LicenceKind.HARD_STOP ? `, ${prisonName}` : ''}`,
```
Comment (andrewrlee) [inline server/services/timelineService.ts:141]: if its hard stop show the PP which prison it came from

**Review by nishanthmahasamudram — DISMISSED** [review]
LGTM

**Review by nishanthmahasamudram — APPROVED** [review]

**Pull request merged**